File: t/CS/PDFDocumentDemo/Digital Signatures/PDFSignaturesHelper.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.Text;

using Leadtools.Pdf;

namespace PDFDocumentDemo
{
   public class PDFSignaturesHelper
   {
      public static string GetSignedByString(string subject)
      {
         if (string.IsNullOrEmpty(subject))
            return "";

         string subjcetInfo = subject;
         if (!subject.EndsWith("/"))
            subjcetInfo += "/";

         string CN = GetSubstring(subjcetInfo, "/CN=", "/");

         string emailAddress = GetSubstring(subjcetInfo, "/emailAddress=", "/");

         string signedBy = CN + " <" + emailAddress + ">";

         return signedBy;
      }

      public static string GetSubjcetOrIssure(string descriptionInfo)
      {
         if (string.IsNullOrEmpty(descriptionInfo))
            return "";

         descriptionInfo += "/";

         string CN = "CN= " + GetSubstring(descriptionInfo, "/CN=", "/") + System.Environment.NewLine;
         string emailAddress = "emailAddress= " + GetSubstring(descriptionInfo, "/emailAddress=", "/") + System.Environment.NewLine;
         string OU = "OU= " + GetSubstring(descriptionInfo, "/OU=", "/") + System.Environment.NewLine;
         string O = "O= " + GetSubstring(descriptionInfo, "/O=", "/") + System.Environment.NewLine;
         string C = "C= " + GetSubstring(descriptionInfo, "/C=", "/");

         descriptionInfo = CN + emailAddress + OU + O + C;

         return descriptionInfo;
      }

      public static string GetKeyUsageString(int KeyUsage)
      {
         string keyUsageString = "";
         string dataSeparator = ", ";

         if ((KeyUsage & PDFSignature.KeyUsageCRLSign) == PDFSignature.KeyUsageCRLSign)
            keyUsageString += "CRL Sign" + dataSeparator;

         if ((KeyUsage & PDFSignature.KeyUsageDataEncipherment) == PDFSignature.KeyUsageDataEncipherment)
            keyUsageString += "Data Encipherment" + dataSeparator;

         if ((KeyUsage & PDFSignature.KeyUsageDecipherOnly) == PDFSignature.KeyUsageDecipherOnly)
            keyUsageString += "Decipher Only" + dataSeparator;

         if ((KeyUsage & PDFSignature.KeyUsageDigitalSignature) == PDFSignature.KeyUsageDigitalSignature)
            keyUsageString += "Digital Signature" + dataSeparator;

         if ((KeyUsage & PDFSignature.KeyUsageEncipherOnly) == PDFSignature.KeyUsageEncipherOnly)
            keyUsageString += "Encipher Only" + dataSeparator;

         if ((KeyUsage & PDFSignature.KeyUsageKeyAgreement) == PDFSignature.KeyUsageKeyAgreement)
            keyUsageString += "Key Agreement" + dataSeparator;

         if ((KeyUsage & PDFSignature.KeyUsageKeyCertSign) == PDFSignature.KeyUsageKeyCertSign)
            keyUsageString += "Key Certificate Sign" + dataSeparator;

         if ((KeyUsage & PDFSignature.KeyUsageKeyEncipherment) == PDFSignature.KeyUsageKeyEncipherment)
            keyUsageString += "Key Encipherment" + dataSeparator;

         if ((KeyUsage & PDFSignature.KeyUsageNonRepudiation) == PDFSignature.KeyUsageNonRepudiation)
            keyUsageString += "Non Repudiation";

         return keyUsageString;
      }

      public static string GetSubstring(string original, string start, string end)
      {
         if (string.IsNullOrEmpty(original))
            return "";

         string substring = "";

         int startPos = 0;
         int length = 0;

         startPos = original.LastIndexOf(start) + start.Length;
         length = original.IndexOf(end, startPos) - startPos;
         if (length > 0)
         {
            substring = original.Substring(startPos, length);
         }
         else
         {
            substring = original.Substring(startPos);
         }

         return substring;
      }

      public static string InsertWhiteSpaceToString(string serialNumber)
      {
         if (string.IsNullOrEmpty(serialNumber))
            return "";

         int count = (serialNumber.Length / 2 - 1);

         for (int i = 0; i < count; i++)
         {
            int index = (2 * i) + (2 + i);
            serialNumber = serialNumber.Insert(index, " ");
         }

         return serialNumber;
      }

      public static string GetState(int state)
      {
         switch (state)
         {
            case PDFSignature.StateInvalid:
               return "INVALID";
            case PDFSignature.StateValid:
               return "VALID";
            default:
               return "UNKNOWN";
         }
      }
   }
}

File: t/VB/PDFDocumentDemo/Digital Signatures/PDFSignaturesHelper.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************

Imports System.Collections.Generic
Imports System.Text

Imports Leadtools.Pdf
Imports System

Namespace PDFDocumentDemo
   Public Class PDFSignaturesHelper
      Public Shared Function GetSignedByString(subject As String) As String
         If String.IsNullOrEmpty(subject) Then
            Return ""
         End If

         Dim subjcetInfo As String = subject
         If Not subject.EndsWith("/") Then
            subjcetInfo += "/"
         End If

         Dim CN As String = GetSubstring(subjcetInfo, "/CN=", "/")

         Dim emailAddress As String = GetSubstring(subjcetInfo, "/emailAddress=", "/")

         Dim signedBy As String = (Convert.ToString(CN & Convert.ToString(" <")) & emailAddress) + ">"

         Return signedBy
      End Function

      Public Shared Function GetSubjcetOrIssure(descriptionInfo As String) As String
         If String.IsNullOrEmpty(descriptionInfo) Then
            Return ""
         End If

         descriptionInfo += "/"

         Dim CN As String = (Convert.ToString("CN= ") & GetSubstring(descriptionInfo, "/CN=", "/")) + System.Environment.NewLine
         Dim emailAddress As String = (Convert.ToString("emailAddress= ") & GetSubstring(descriptionInfo, "/emailAddress=", "/")) + System.Environment.NewLine
         Dim OU As String = (Convert.ToString("OU= ") & GetSubstring(descriptionInfo, "/OU=", "/")) + System.Environment.NewLine
         Dim O As String = (Convert.ToString("O= ") & GetSubstring(descriptionInfo, "/O=", "/")) + System.Environment.NewLine
         Dim C As String = Convert.ToString("C= ") & GetSubstring(descriptionInfo, "/C=", "/")

         descriptionInfo = Convert.ToString(Convert.ToString(Convert.ToString(CN & emailAddress) & OU) & O) & C

         Return descriptionInfo
      End Function

      Public Shared Function GetKeyUsageString(KeyUsage As Integer) As String
         Dim keyUsageString As String = ""
         Dim dataSeparator As String = ", "

         If (KeyUsage And PDFSignature.KeyUsageCRLSign) = PDFSignature.KeyUsageCRLSign Then
            keyUsageString += Convert.ToString("CRL Sign") & dataSeparator
         End If

         If (KeyUsage And PDFSignature.KeyUsageDataEncipherment) = PDFSignature.KeyUsageDataEncipherment Then
            keyUsageString += Convert.ToString("Data Encipherment") & dataSeparator
         End If

         If (KeyUsage And PDFSignature.KeyUsageDecipherOnly) = PDFSignature.KeyUsageDecipherOnly Then
            keyUsageString += Convert.ToString("Decipher Only") & dataSeparator
         End If

         If (KeyUsage And PDFSignature.KeyUsageDigitalSignature) = PDFSignature.KeyUsageDigitalSignature Then
            keyUsageString += Convert.ToString("Digital Signature") & dataSeparator
         End If

         If (KeyUsage And PDFSignature.KeyUsageEncipherOnly) = PDFSignature.KeyUsageEncipherOnly Then
            keyUsageString += Convert.ToString("Encipher Only") & dataSeparator
         End If

         If (KeyUsage And PDFSignature.KeyUsageKeyAgreement) = PDFSignature.KeyUsageKeyAgreement Then
            keyUsageString += Convert.ToString("Key Agreement") & dataSeparator
         End If

         If (KeyUsage And PDFSignature.KeyUsageKeyCertSign) = PDFSignature.KeyUsageKeyCertSign Then
            keyUsageString += Convert.ToString("Key Certificate Sign") & dataSeparator
         End If

         If (KeyUsage And PDFSignature.KeyUsageKeyEncipherment) = PDFSignature.KeyUsageKeyEncipherment Then
            keyUsageString += Convert.ToString("Key Encipherment") & dataSeparator
         End If

         If (KeyUsage And PDFSignature.KeyUsageNonRepudiation) = PDFSignature.KeyUsageNonRepudiation Then
            keyUsageString += "Non Repudiation"
         End If

         Return keyUsageString
      End Function

      Public Shared Function GetSubstring(original As String, start As String, [end] As String) As String
         If String.IsNullOrEmpty(original) Then
            Return ""
         End If

         Dim substring As String = ""

         Dim startPos As Integer = 0
         Dim length As Integer = 0

         startPos = original.LastIndexOf(start) + start.Length
         length = original.IndexOf([end], startPos) - startPos
         If length > 0 Then
            substring = original.Substring(startPos, length)
         Else
            substring = original.Substring(startPos)
         End If

         Return substring
      End Function

      Public Shared Function InsertWhiteSpaceToString(serialNumber As String) As String
         If String.IsNullOrEmpty(serialNumber) Then
            Return ""
         End If

         Dim count As Integer = CInt(serialNumber.Length / 2 - 1)

         For i As Integer = 0 To count - 1
            Dim index As Integer = (2 * i) + (2 + i)
            serialNumber = serialNumber.Insert(index, " ")
         Next

         Return serialNumber
      End Function

      Public Shared Function GetState(state As Integer) As String
         Select Case state
            Case PDFSignature.StateInvalid
               Return "INVALID"
            Case PDFSignature.StateValid
               Return "VALID"
            Case Else
               Return "UNKNOWN"
         End Select
      End Function
   End Class
End Namespace
